### PR TITLE
Started implementing Telegram bot

### PR DESCRIPTION
**Pull request by jonathanschneider**
So I was wondering how BierBot could let a user know that the brew is finished or user action is required when the user isn't within hearing distance. I thought a Telegram bot might be a great idea so I started implementing it.

What I've done so far:
- Created a script to send messages to Telegram
- Called the script when the brew is finished
- Added necessary parameters to the Mongoose schema and the default settings
- Adjusted package.json accordingly

If you're interested, this is still missing because I haven't quite figured out completely how everything plays together:
- Add fields to the settings page for the token, the chat ID and to enable/disable Telegram and write to data base
- Get token and chat ID from data base instead of constants
- Call the script when the next step in the recipe requires user action
- Condition to only call the script if Telegram is enabled

Since this is not yet function you might not want to merge it into your master branch. If you're interested and want to create a develop branch I will create another pull request.

### CODE REVIEW

**Comment by BernhardSchlegel** [comment]
Haha, how awsome is that?! I created a develop branch, but feel free to commit to your PR and we merge straight into master as soon as it's ready. Here are some entry points for your questions:

1. Add fields to settings page: have a look at [Settings.html](https://github.com/BernhardSchlegel/BierBot/blob/master/client/partials/settings.html), you can pretty much copy & paste from a different section. The `settings` object itself is transmitted "as whole" to the nodeJS backend (no need to adjust the socket). The logic resides in the (way too big, I know...) [app.js](https://github.com/BernhardSchlegel/BierBot/blob/master/client/js/app.js) file, starting line 1218.
2. Storing resolves pretty much along with the first issue / question. As soon as the `token`, `chatID`, and `telegram_enabled`-state are fields of the settings object, storing will work just fine (it's noSQL mongo db). Just do not forget to add a field in the mongoose object [appsettings.js](https://github.com/BernhardSchlegel/BierBot/blob/master/server/app/models/appsetting.js)
3. That's easy, there is one global option in in server.js, called [setNextStep](https://github.com/BernhardSchlegel/BierBot/blob/21be6d459396d16fac1da98a0a736b80f9a0dc62/server/server.js#L2114), integrate it somewhere there, but ... 
4. do not forget to condition on `appSettings.telegram_enabled` (as you said :) )

Feel free to reach out again, anytime!

Cheers

**Comment by jonathanschneider** [comment]
@BernhardSchlegel any idea if updating the request package to v2.88.2 will break something?

**Comment by jonathanschneider** [comment]
Screenshot 2021-01-13 at 20 39 43

**Comment by BernhardSchlegel** [comment]
> @BernhardSchlegel any idea if updating the request package to v2.88.2 will break something?

There is a fair chance that it will work since the major version stays the same. Give it a try and let me know. I'm happy to help with debugging if something goes sideways...

**Comment by jonathanschneider** [comment]
Okay, I added the required settings to the front end, created the model and set default settings. I am not sure how to get the settings from the data base in telegram.js.
I wanted the Telegram implementation in its own file even though it isn't much code. Since all settings are already available in server.js would it make more sense to implement Telegram there than in its own file?
Could you have a look?

**Comment by BernhardSchlegel** [comment]
Sure, I'm happy to! Putting it in a separate file is cleaner way to go - even if it's just a few lines.

How I would do it:

- Import (=`require`) the Settings object (e.g. as done in [pd.js](https://github.com/jonathanschneider/BierBot/blob/dcf0de2b8723f3cb9633b30920eb8dcabc3152aa/server/libs/pd.js#L7)) into [telegram.js](https://github.com/jonathanschneider/BierBot/blob/telegram/server/libs/telegram.js#L1).
- Move `bot` below the imports, make it a module wide available `var` (not a `const`). Delete variable `token` and `chatId` - these should be part of your settings object.
- Create an exported method called `init` In [telegram.js](https://github.com/jonathanschneider/BierBot/blob/telegram/server/libs/telegram.js#L1). This will init the bot (`bot = new TelegramBot(token, {polling: false});`) using the token from Settings, and set the `chatId`. This function should be called on startup after the settings are loaded (e.g. [here](https://github.com/jonathanschneider/BierBot/blob/dcf0de2b8723f3cb9633b30920eb8dcabc3152aa/server/server.js#L702)) and whenever the settings are changed ([here](https://github.com/jonathanschneider/BierBot/blob/dcf0de2b8723f3cb9633b30920eb8dcabc3152aa/server/server.js#L1059)).

*Backup*: If importing the Settings module in `telegram.js` won't work, you can pass the a `settings` object to the `telegram::init()` function on every call. In this case, you cannot delete the `token` and `chatId` variables.

File: server/libs/defaults.js [hunk 1]
```diff
@@ -18,6 +18,11 @@ var defaultSettings = {
    "wlanPassphrase": "deinPW",
    "wlanSSID": "unbekanntes WLAN",
    "motorWarningChecked": false,
-   "addToSensorVal": 0
+   "addToSensorVal": 0,
```
Comment (BernhardSchlegel) [inline server/libs/defaults.js:21]: @jonathanschneider Does storing and object of objects work? I'm having a hard time to recap, but there might have been a reason that made me keep the settings object flat. Workaround camelCase: `telegramEnabled`, `telegramToken`, `telegramChatId`

**Comment by jonathanschneider** [comment]
Defining an object in a schema is possible according to https://mongoosejs.com/docs/guide.html but if the storing in settings.html works I don't know yet. Haven't gotten that far to try to access the object.

File: server/libs/defaults.js (hunk 1 above)
Comment (jonathanschneider) [reply server/libs/defaults.js:21]: Defining an object in a schema is possible according to https://mongoosejs.com/docs/guide.html but if the storing in settings.html works I don't know yet. Haven't gotten that far to try to access the object.

**Comment by jonathanschneider** [comment]
Thanks for the hints and the idea with the init function @BernhardSchlegel. That was really helpful.
I managed to access the settings in telegram.js but after implementing the init method, I decided to just go with your backup plan. Wherever the init method is called, the settings are "right there", so why not just pass them as arguments?

I believe I am done with this enhancement then. Check it out.

File: server/libs/defaults.js (hunk 1 above)
Comment (jonathanschneider) [reply server/libs/defaults.js:21]: Yup, it works!

**Comment by BernhardSchlegel** [comment]
I'll have a look at the weekend !

File: server/server.js [hunk 2]
```diff
@@ -633,15 +633,15 @@ var heatCool = function(targetState) {
...
+};
 var safeModeHeatingCoolingWarned = false;
 var setHeatingCooling = function(targetState) {
    if (heatingCoolingResetMode == true) {
       brewlog.log("skipping setHeatingCooling since heatingCoolingResetMode=true and force=false...");
       return;
    }
    if (safeModeActive == true) {
-      if (safeModeHeatingCoolingWarned = false) {
+      if (safeModeHeatingCoolingWarned == false) {
```
Comment (BernhardSchlegel) [inline server/server.js:644]: Oh Lord... Thanks for that

Comment (jonathanschneider) [reply server/server.js:644]: It happens 😉

**Comment by jonathanschneider** [comment]
So, I brewed my first beer with BierBot yesterday and the Telegram messages came in really handy. Since currently messages are only sent when BierBot enters a manual step I added a few more manual steps to my recipe just to go the reminders. I now think it’d be better if BierBot sent messages on every step transition.
What do you think?

**Comment by BernhardSchlegel** [comment]
Hi Jonathan! I think that's a great idea. Can you open a new PR for this (I merged this one as of today)? Also, as I tested the telegram integration today I noticed two things: 1st: it worked pretty great out of the box 2nd: The option to provide multiple chat ids would come in super handy (we are a group of 4 hobby brewers). A simple way to implement it would be to allow CSV in the SettingsPage (`chatID1,chatID2`) and then do a simple `settingsObject.chatID.split(',')` to iterate over all chat ids (nice side effect of split is, that if there is no `,` at all in the string, it returns (still) an array of size 1). Let me know what you think!

**Comment by jonathanschneider** [comment]
I’ll look into it. In the meantime you could also create a Telegram group between the four of you (and IDBot) and provide the group‘s chat ID.

**Comment by jonathanschneider** [comment]
Plus the bot you created of course!

**Comment by BernhardSchlegel** [comment]
ahhh ! Didn't know that, my feature request does not make soo much sense than after all

**Comment by jonathanschneider** [comment]
I don’t know. With your suggestion you could just change the chat IDs for the brewers present on that day without having to edit the Telegram group. Since it has no disadvantage for single chat IDs, why not? I’ll look into it.

**Comment by BernhardSchlegel** [comment]
I appreciate that!

**Pull request merged**